### ISSUE TITLE
Account for possible cache concurrency issues with file systems not supporting atomic moves.

### ISSUE THREAD

**Issue by coollog** [GoogleContainerTools/jib#1044]
From https://github.com/GoogleContainerTools/jib/pull/1037#discussion_r220692899

For file systems that don't support atomic move, the cache is not guaranteed to be in a fully valid state between reads. We probably have to have an additional locking mechanism for such file systems.

**Comment by chanseokoh** [comment]
Just FTR, I've looked into the `Files.move()` code, and as soon as the source file system and the move target file system are different, `StandardCopyOption.ATOMIC_MOVE` will always fail (throwing an exception). Also in this case, `move()` will always degrade into copy+delete without `ATOMIC_MOVE`.

**Comment by chanseokoh** [comment]
Just wanted to make see some actual errors in practice, so I tried the following.

When using `openjdk:8` as a base image, one of the directories created in the cache directory is `jib/layers/60c0e52d1ec297a9ade35f6b17bcf69d959bb1f16c506290f29af5513b0cdabd`. In `DefaultCacheStorageWrite`, I put the following custom code in the `DirectoryNotEmtpyException` block of `moveIfDoesNotExist()`, so that the copy is blocked just before copying files into that directory (busy waiting). The code can proceed only after seeing a file `/tmp/signal`. With this, I can artificially synchronize two concurrent Jib builds to run the same code section almost at the same time.
 
 ```java
      } catch (DirectoryNotEmptyException ex) {
        if (destination
            .getFileName()
            .toString()
            .equals("60c0e52d1ec297a9ade35f6b17bcf69d959bb1f16c506290f29af5513b0cdabd"))) {
          while (true) { // busy wait
            try (InputStream in = Files.newInputStream(Paths.get("/tmp/signal")) {
              break;
            } catch (IOException e) {
            }
          }
        }
```

Then, we can get `FileAlreadyExistsException` from `Files.createDirectories()` and `Files.copy()`. But note that if Jib degrades into copying files, the cache briefly goes into in an incomplete state anyway, as copying files is not an atomic operation. However, in practice I think it will be extremely unlikely that someone will hit a problem from this issue, though theoretically we have the possibility.

**Comment by chanseokoh** [comment]
> But note that if Jib degrades into copying files, the cache briefly goes into in an incomplete state anyway, as copying files is not an atomic operation. However, in practice I think it will be extremely unlikely that someone will hit a problem from this issue, though theoretically we have the possibility.

https://github.com/GoogleContainerTools/jib/issues/1815#issuecomment-577886257 might be an example of a brief incomplete state.

**Comment by loosebazooka** [comment]
I believe this is handled in code now (as much as we possibly can)? @chanseokoh? should we close

**Comment by chanseokoh** [comment]
This is still an issue in that Jib has a concurrency issue if atomic move is not supported by the filesystem. We can definitely eliminate the concurrency issue (for example, using a lock file as we do with manifest caching). But this has been a very low priority, and especially we did not receive issue reports around this for some time. Perhaps almost all the time, atomic move is supported in practice, or the user just retries (possibly after deleting our cache). Given the priority, we might never a chance to get back to this issue. I'm fine closing this.
